src/Main.elm

```diff
@@ -1 +1,313 @@
-a=0
+module Main exposing (Model, Msg, init, subscriptions, update, view)
+
+import Browser
+import Html exposing (Html, div, p, text)
+import Http exposing (Error)
+import Json.Decode exposing (Decoder, field, int, list, map2, string)
+
+
+
+-- APP
+
+
+main : Program () Model Msg
+main =
+    Browser.element
+        { init = init
+        , view = view
+        , update = update
+        , subscriptions = subscriptions
+        }
+
+
+init : () -> ( Model, Cmd Msg )
+init _ =
+    ( Model Loading Loading Maybe.Nothing "code"
+    , Cmd.batch [ getRhymes "code", getPhrases ]
+    )
+
+
+
+-- MODEL
+
+
+type alias Model =
+    { rhymes : RhymeResult
+    , phrases : PhraseResult
+    , puns : Maybe (List Pun)
+    , word : String
+    }
+
+
+type RemoteData a
+    = Loading
+    | Failure
+    | Success a
+
+
+type alias Rhyme =
+    { word : String
+    , score : Int
+    }
+
+
+type Phrase
+    = Phrase String
+
+
+type alias RhymeResult =
+    RemoteData (List Rhyme)
+
+
+type alias PhraseResult =
+    RemoteData (List Phrase)
+
+
+type Pun
+    = Pun String
+
+
+
+-- UPDATE
+
+
+type Msg
+    = GotRhymes (Result Error (List Rhyme))
+    | GotPhrases (Result Error String)
+    | SetPuns Model
+
+
+update : Msg -> Model -> ( Model, Cmd Msg )
+update msg model =
+    case msg of
+        GotRhymes result ->
+            case result of
+                Ok rhymes ->
+                    let
+                        newModel =
+                            { model | rhymes = Success (goodRhymes rhymes) }
+                    in
+                    update (SetPuns newModel) newModel
```

Comment:
This does seem a little roundabout. So it looks like `SetPuns` message results in `punList` getting called to actually put the puns together. What if you updated `punList` to expect situations where `model.rhymes` or `model.phrases` aren't actually loaded yet? They're wrapped as a `RemoteData` type, so you can add some sort of case statement that checks the status of each. Something like:

```
punList : Model -> List Pun
punList model =
    case (model.phrases, model.rhymes) of
        (Loading, _) ->
           // return some incomplete result?
        (_, Loading) ->
           // ditto
        (Success , Success) ->
               let
                   phrases =
                       phraseResultToPhraseList model.phrases

                   rhymes =
                       rhymeList model.rhymes
               in
               List.filter (hasMatchingRhymes rhymes) phrases
               |> List.map
                   (rhymesAndWordMakePhrasePunchline rhymes model.word)
               |> List.map Pun
```

This way you can call `punList` directly after setting `model.rhymes` or `model.phrases` when receiving `GotRhymes` and `GotPhrases`. No need for the extra update call with `SetPuns`. Does that work?

Reply:
Yeah, that does feel weird. Here are a few thoughts:

Recursively calling `update` from within itself is considered bad practice. If you want to share logic across multiple branches, the usual solution is to extract a function.

Reply:
Your modeling has two sources of truth for the puns: `model.puns` and a combination of `model.phrases` and `model.rhymes`. It allows the two sources of truth to get out of sync, allowing states like:

```elm
{ phrases = Loading
  , rhymes = Loading
  , puns = Just ["pun1", "pun2"]
  , word = "pun"
}
```

I'm guessing such a state is invalid? You might want to eliminate `model.puns` entirely in favor of calling the `punList` function directly any time you need puns. That way you know your puns always reflect the current data.

Reply:
Finally, I'm curious about the lifecycle of phrases and rhymes. Are they both always fetched together? Is one or the other only fetched at boot time? There are a few different tools at your disposal for handling these situations including different approaches to structuring your `Model`, and the `Task` API which works kind of like promises in JS.

Reply:
Thanks both for the detailed feedback! You're totally right @joelq that is an invalid state at the moment. It's possible that eventually it won't be (multiple phrase lists and different pun api calls) but it's not at the moment.

Currently they are always fetched together because they are both effectively static. It's always the same list of rhymes and phrases. And I know I keep saying it, but eventually they will be much more separated, but this is how I figured out how to kick off two tasks at once. 100% open to other ideas 😄 

For now, I'm going to try and just call the `punList` method in the view!